File: Day3/AzureModel.vb
Imports System.ComponentModel
Imports System.Runtime.CompilerServices
Imports Microsoft.WindowsAzure.MobileServices

Namespace Models
    Public Class TodoItem
        Public Property Id As String
        Public Property UserId As String
        Public Property Hand As Integer
        Public Result As Integer
    End Class

    Public Class UserItem
        Public Property Id As String
        Public UserId As String
        Public UserName As String
    End Class


    Public Class AzureModel
        Implements INotifyPropertyChanged

        Private Shared MobileServiceClient As New MobileServiceClient("https://leapcloud.azure-mobile.net")

        Private Items As MobileServiceCollection(Of TodoItem, TodoItem)
        Private TodoTable As IMobileServiceTable(Of TodoItem) = MobileServiceClient.GetTable(Of TodoItem)()
        Private UserTable As IMobileServiceTable(Of UserItem) = MobileServiceClient.GetTable(Of UserItem)()

        Private _UserId As String = ""
        Public Property UserId As String
            Get
                Return Me._UserId
            End Get
            Set(value As String)
                Me._UserId = value
                OnPropertyChanged()
            End Set
        End Property

        '-1:受付中、0:対戦中、1:勝利、2:敗北
        Private _Result As Integer = -1
        Public Property Result As Integer
            Get
                Return Me._UserId
            End Get
            Set(value As Integer)
                Me._Result = value
                OnPropertyChanged()
            End Set
        End Property

        Private _ErrorString As String
        Public Property ErrorString As String
            Get
                Return Me._ErrorString
            End Get
            Set(value As String)
                Me._ErrorString = value
                OnPropertyChanged()
            End Set
        End Property

        Public Async Function SetUserId(userName As String) As Task
            Try
                Await GetUserID(userName)
                If Me.UserId.Length = 0 Then
                    Await UserTable.InsertAsync(New UserItem With {.UserName = userName})
                    Await GetUserID(userName)
                End If
            Catch ex As Exception
                Me.ErrorString = ex.Message
            End Try
        End Function

        Public Async Function GetUserID(userName As String) As Task
            Dim query = UserTable.CreateQuery()
            query.Parameters.Add("UserName", userName)
            Dim response = Await query.ToListAsync()
            If response.Count > 0 Then
                Me.UserId = response(0).UserId
            Else
                Me.UserId = ""
            End If
        End Function

        Public Async Function SetItem(hand As Integer) As Task
            Try
                Await TodoTable.InsertAsync(New TodoItem With {.UserId = Me.UserId, .Hand = hand})
                Await GetItem(hand)
            Catch ex As Exception
                Me.ErrorString = ex.Message
            End Try
        End Function

        Public Async Function GetItem(hand As Integer) As Task
            Dim query = TodoTable.CreateQuery()
            query.Parameters.Add("UserId", UserId)
            Dim response = Await query.ToListAsync()
            If response.Count > 0 Then
                Me.Result = response(0).Result
            Else
                Me.Result = -1
            End If
        End Function

        Public Event PropertyChanged(sender As Object,
                                     e As PropertyChangedEventArgs) Implements INotifyPropertyChanged.PropertyChanged
        Protected Sub OnPropertyChanged(<CallerMemberName> Optional propertyName As String = Nothing)
            RaiseEvent PropertyChanged(Me, New PropertyChangedEventArgs(propertyName))
        End Sub
    End Class
End Namespace

File: Day3/AzureModel.cs
using Microsoft.WindowsAzure.MobileServices;
using System;
using System.ComponentModel;
using System.Runtime.CompilerServices;
using System.Threading.Tasks;

namespace leapcloud
{
    namespace Model
    {
        public class TodoItem
        {
            public string Id { get; set; }
            public string UserId { get; set; }
            public int Hand { get; set; }
            public int Result { get; set; }
        }

        public class UserItem
        {
            public string id { get; set; }
            public string UserId { get; set; }
            public string UserName { get; set; }
        }

        public class AzureModel : INotifyPropertyChanged
        {
            private static MobileServiceClient MobileServiceClient = new MobileServiceClient("https://leapcloud.azure-mobile.net");
            private MobileServiceCollection<TodoItem, TodoItem> Items;
            private IMobileServiceTable<TodoItem> TodoTable = MobileServiceClient.GetTable<TodoItem>();
            private IMobileServiceTable<UserItem> UserTable = MobileServiceClient.GetTable<UserItem>();

            private string _UserId = "";
            public string UserId
            {
                get { return this._UserId; }
                set
                {
                    this._UserId = value;
                    OnPropertyChanged();
                }
            }

            //'-1:受付中、0:対戦中、1:勝利、2:敗北
            private int _Result = -1;
            public int Result
            {
                get { return this._Result; }
                set
                {
                    this._Result = value;
                    OnPropertyChanged();
                }
            }

            private string _ErrorString = "";
            public string ErrorString
            {
                get { return this._ErrorString; }
                set
                {
                    this._ErrorString = value;
                    OnPropertyChanged();
                }
            }

            public async Task SetUserId(string userName)
            {
                try
                {
                    await GetUserID(userName);
                    if (this.UserId.Length == 0)
                    {
                        await UserTable.InsertAsync(new UserItem {UserName = userName});
                        await GetUserID(userName);
                    }
                }
                catch (Exception ex)
                {
                    this.ErrorString = ex.Message;
                }
            }

            public async Task GetUserID(string userName)
            {
                var query = UserTable.CreateQuery();
                query.Parameters.Add("UserName",userName);

                var response = await query.ToListAsync();
                if (response.Count > 0)
                {
                    this.UserId = response[0].UserId;
                }
                else
                {
                    this.UserId = "";
                }
            }

            public async Task SetItem(int hand)
            {
                try
                {
                    await TodoTable.InsertAsync(new TodoItem {UserId = this.UserId, Hand = hand});
                    await GetItem(hand);
                }
                catch(Exception ex)
                {
                    this.ErrorString = ex.Message;
                }

            }

            public async Task GetItem(int hand)
            {
                var query = TodoTable.CreateQuery();
                query.Parameters.Add("UserId", UserId);
                var response = await query.ToListAsync();
                if (response.Count > 0)
                {
                    this.Result = response[0].Result;
                }
                else
                {
                    this.Result = -1;
                }
            }

            public event PropertyChangedEventHandler PropertyChanged;
            protected virtual void OnPropertyChanged([CallerMemberName] String propertyName = "")
            {
                var handler = this.PropertyChanged;
                if (handler != null)
                    handler(this, new PropertyChangedEventArgs(propertyName));
            }
        }        
    }
}
